### PR TITLE
Add support for the Cart & Checkout in WooCommerce's declare compatibility feature

### LINKED ISSUE #36413: [Enhancement]: Add support for the Cart & Checkout Blocks in the declaring compatibility feature
### Context

In Rukib team, we are maintaining the [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks) plugin! Couple of months ago we did a research spike (pdFofs-mB-p2) about managing extensions that don't support the Cart & Checkout Blocks. One of our propositions was to declare (in)compatibility with the C&C Blocks (pdFofs-mB-p2#declare-in-compatibility-with-the-cart-checkout-blocks) using the same approach in [HPOS compatibility declaration](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#declaring-extension-incompatibility). During one of our pair programming and brainstorming sessions (pca54o-4NP-p2#declaring-in-compatibly-with-the-c-c-blocks) we worked in creating a prouf of concept for declaring (in)compatibility with the Cart & Checkout Blocks feature.

Declaring (in)compatibility is simple:
- Declare WC tested up to in the main plugin file(s)
- Add the correct code snippet to the main plugin file(s)

The code snippet used for declaring (in)compatibility with [HPOS](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#declaring-extension-incompatibility):

```PHP
add_action( 'before_woocommerce_init', function() {
    if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
        \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
    }
} );
```

So, the function of interest is: `declare_compatibility( 'custom_order_tables', FILE, true );`, see [function definition](https://href.li/?https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Utilities/FeaturesUtil.php#L45-L67):

https://github.com/woocommerce/woocommerce/blob/c3d62f113addbbf820ed5e8103a0a052092efc00/plugins/woocommerce/src/Utilities/FeaturesUtil.php#L45-L67

In theory, we want to replace the `custom_order_tables` feature_id by a C&C Blocks feature_id (e.g., `cart_checkout_blocks`). During our research, we realized that only the WooCommerce features in `WooCommerce -> Settings -> Advanced -> Features` can use the `declare_compatibility` function:

image

In other words, the current implementation of the `declare_compatibility` function doesn’t work for the Cart & Checkout Blocks feature.

### Solutions

I this case, we have two options:

- In Woo Blocks, create the compatibility declaration API from scratch. This, I believe, is a bad idea: we will duplicate the same feature, and we will not have a single source of truth!
- Add support for the Cart & Checkout Blocks in the declaring compatibility feature

We eventually chose to go with the second solution!

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tarhi-saad**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

For a more detailed description, please read the one in this ticket #36413!

In summary, we added the Cart & Checkout Blocks to the list of features so that other extension developers might declare (in)compatibility with it. For the time being, we are preventing the display of the Cart & Checkout Blocks feature's UI settings. We may choose otherwise in the future, but for the time being, we only need the Cart & Checkout Blocks to be included in the list of WooCommerce features.

Closes #36413.

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

Checkout to this branch, then:

1. Create a new plugin or clone an existing one, install it, and activate it on your test website
2. Declare its compatibility with the Cart & Checkout Blocks by adding the following snippet into the main plugin file:

```PHP

add_action( 'before_woocommerce_init', function() {
	if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
		\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'cart_checkout_blocks', __FILE__, true );
	}
} );

```

3. Clone the [WooCommerce Blocks](https://github.com/woocommerce/woocommerce-blocks) plugin, install it, and activate it in your test website.
4. Check if we can get the list of compatible plugins with the Cart & Checkout Blocks with the WooCommerce Blocks plugin, and that the name of the plugin from step 1 is included. We can use the function `get_compatible_plugins_for_feature` from the `FeaturesController` API. For example, we can add the following snippet within the main plugin file:

```PHP

add_action(
	'woocommerce_init',
	function() {
		if ( class_exists( '\Automattic\WooCommerce\Utilities\FeaturesUtil' ) ) {
			var_dump(
				\Automattic\WooCommerce\Utilities\FeaturesUtil::get_compatible_plugins_for_feature( 'cart_checkout_blocks' )
			);
		}
	}
);

```

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (opr) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php]: Isn't `cart_checkout_blocks` a key of `$features` already, meaning `$feature_ids` would include it? Is there a reason you had to specifically add it here?

**Review by opr — COMMENTED** [review]
Hey Saad, thanks for working on this. The functionality is working fine! Just one comment that needs to be resolved from my end.

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (tarhi-saad) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php]: Hi @opr! Thank you for the review! 🙌 

> Isn't cart_checkout_blocks a key of $features already, meaning $feature_ids would include it? Is there a reason you had to specifically add it here?

In this line, we are removing the `cart_checkout_blocks` to exclude it from the `mature_feature_ids` list. We don't want the add a `setting UI` for the Cart & Checkout feature. We will get the following result without this line:

image

Comment (opr) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php]: Great, thanks for explaining!

**Review by opr — DISMISSED** [review]

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php]: While this works for what you intend it to, I feel we should have a more generic approach to adding features that don't need (or want) to expose a setting in the UI. For example, similar to the `is_experimental` flag we could have a `disable_ui` flag or something like that and then we would exclude those settings from `$feature_ids` below.

I'd be more than happy to review that change if you want to include it in this PR. If not, that's also fine, and we can add it later via a different PR.

Let me know what your thoughts are.

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php]: The arrows here are not correctly aligned resulting in a PHPCS error: https://github.com/woocommerce/woocommerce/actions/runs/3928725851/jobs/6716673300#step:6:20.

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @tarhi-saad!

Overall, this looks good, though there's a tiny PHPCS issue that needs to be fixed. I also left some comments re: the approach to hiding the UI, but that's not necessarily a blocker. Please let me know what you think!

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (tarhi-saad) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php]: This is an awesome suggestion! 💯 Thank you, @jorgeatorres! 🙌

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (tarhi-saad) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php]: Nice catch! 👍

It's weird! Running PHP lint locally shows me different results than the ones detected in CI.

**Comment by tarhi-saad** [comment]
Hello, @jorgeatorres! 👋 Thank you for the review. And I apologize for addressing your feedback this late. Last week I was release manager, and I couldn't find the time for my ongoing tasks! Please could you give it another review?

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php]: We should not exclude features with disabled UI here as `$mature_feature_ids` as that wouldn't prevent experimental features with `disable_ui = true` from having their settings rendered.

My personal suggestion, would be to add a check on [this line](https://github.com/woocommerce/woocommerce/blob/615365d14e0e2b944f05213e366fd1a9849612db/plugins/woocommerce/src/Internal/Features/FeaturesController.php#L595) such as

```php
if ( isset( $features[ $id ]['disable_ui'] ) && $features[ $id ]['disable_ui'] ) {
	continue;
}
```

That way, we can skip displaying the setting for features with disabled UI whether it is a mature or experimental feature.

As a side effect, we can get rid of the `$disabled_ui_feature_ids` var, at least for the time being.

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @tarhi-saad!

Thanks for making the changes. I noticed a tiny issue with the way `disable_ui` filtering was implemented, and so I left a suggestion in the code. Other than that, this is looking great 💯.

Please let me know what you think.

**Review by jorgeatorres — APPROVED** [review]
@tarhi-saad: This looks good to me. Thanks for making the changes! 💯

I see you requested a review from others, so I'm not sure if you want to wait until they review. If not, feel free to go ahead with the merging part (or ping me if you're not allowed to do that).

**Comment by tarhi-saad** [comment]
Thank you, @jorgeatorres, for your review and the PR approval! 🙌 Your feedback was incredibly valuable. Thanks to your suggestions, the code is now looking better than ever! I'll go ahead and merge it!

**Pull request merged**
